src/lib/parserRaw.c

```diff
@@ -583,6 +583,119 @@ RdbStatus elementRawSet(RdbParser *p) {
+
+        /* Following enums are aligned to module-opcodes. Static assert below. */
+        ST_RAW_MODULE_OPCODE_SINT,
+        ST_RAW_MODULE_OPCODE_UINT,
+        ST_RAW_MODULE_OPCODE_FLOAT,
+        ST_RAW_MODULE_OPCODE_DOUBLE,
+        ST_RAW_MODULE_OPCODE_STRING_CALL_STR,
+
+        ST_RAW_MODULE_OPCODE_STRING_STR_RETURN, /* return from sub-element string parsing */
+        ST_RAW_MODULE_NEXT_OPCODE,
+    } RAW_MODULE_STATES;
+
+    RawContext *rawCtx = &p->rawCtx;
+
+    while (1) {
+        switch (p->elmCtx.state) {
+
+            case ST_RAW_MODULE_GET_ID: {
+                int len = 0;
+                uint64_t moduleid = 0;
+
+                aggMakeRoom(p, 32);
+                IF_NOT_OK_RETURN(rdbLoadLen(p, NULL, &moduleid, (unsigned char *) rawCtx->at, &len));
+                /*** ENTER SAFE STATE ***/
+                IF_NOT_OK_RETURN(cbHandleBegin(p, DATA_SIZE_UNKNOWN_AHEAD));
+                IF_NOT_OK_RETURN(aggUpdateWritten(p, len));
+
+                updateElementState(p, ST_RAW_MODULE_NEXT_OPCODE);
+                break;
+            }
+
+            case ST_RAW_MODULE_OPCODE_SINT:
+            case ST_RAW_MODULE_OPCODE_UINT: {
+                uint64_t val = 0;
+                int len = 0;
+
+                IF_NOT_OK_RETURN(aggMakeRoom(p, 32));
+                IF_NOT_OK_RETURN(rdbLoadLen(p, NULL, &val, (unsigned char *) rawCtx->at, &len));
+                /*** ENTER SAFE STATE ***/
+                IF_NOT_OK_RETURN(aggUpdateWritten(p, len));
+                updateElementState(p, ST_RAW_MODULE_NEXT_OPCODE);
+                break;
+            }
+
+            case ST_RAW_MODULE_OPCODE_FLOAT: {
+                IF_NOT_OK_RETURN(aggMakeRoom(p, sizeof(float)));
+                IF_NOT_OK_RETURN(rdbLoadFloatValue(p, (float *) rawCtx->at));
+                /*** ENTER SAFE STATE ***/
+                IF_NOT_OK_RETURN(aggUpdateWritten(p, sizeof(float)));
+                updateElementState(p, ST_RAW_MODULE_NEXT_OPCODE);
+                break;
+            }
+
+            case ST_RAW_MODULE_OPCODE_DOUBLE: {
+                IF_NOT_OK_RETURN(aggMakeRoom(p, sizeof(double)));
+                IF_NOT_OK_RETURN(rdbLoadDoubleValue(p, (double *) rawCtx->at));
+                /*** ENTER SAFE STATE ***/
+                IF_NOT_OK_RETURN(aggUpdateWritten(p, sizeof(double)));
+                updateElementState(p, ST_RAW_MODULE_NEXT_OPCODE);
+                break;
+            }
+
+            case ST_RAW_MODULE_OPCODE_STRING_CALL_STR: {
+                /* call raw string as subelement */
+                return subElementCall(p, PE_RAW_STRING, ST_RAW_MODULE_OPCODE_STRING_STR_RETURN);
+            }
+
+            case ST_RAW_MODULE_OPCODE_STRING_STR_RETURN: {
+                /*** ENTER SAFE STATE (no rdb read)***/
+                size_t len;
+                char *dataRet;
+                subElementCallEnd(p, &dataRet, &len);
+                updateElementState(p, ST_RAW_MODULE_NEXT_OPCODE);
+                break;
```

Comment:
what's that safe state? safe for what and when does it end?

Reply:
Available for a while and exaplained in the README. Please see section `Parsing-Element states` and subsection ` Defining Safe-state`. Please let me know if it is clear enough.